### PR TITLE
new ui layout for project navigation

### PR DESCRIPTION
**Pull request by natibekele**
new project nav control
new  chapter and verse cards

### CODE REVIEW

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditorStyles.kt [hunk 1]
```diff
@@ -52,14 +71,89 @@ class ProjectEditorStyles : Stylesheet() {
...
+            prefHeight = 210.px
+            maxWidth = 180.px
+            maxHeight = 210.px
+            backgroundColor += Color.WHITE
+            backgroundRadius += box(5.0.px)
+            padding = box(0.0.px, 0.0.px,10.0.px, 0.0.px)
+            effect = DropShadow(2.0, 4.0, 6.0, Color.LIGHTGRAY)
+
+            cardBackground {
+                backgroundColor += c("#E6E8E9")
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditorStyles.kt]: can we put this color somewhere?

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditorStyles.kt [hunk 2]
```diff
@@ -52,14 +71,89 @@ class ProjectEditorStyles : Stylesheet() {
...
+            }
+            alignment = Pos.CENTER
+
+            chaptercardGraphic{
+                prefWidth = 160.0.px
+                prefHeight = 140.14.px
+                maxWidth = 160.0.px
+                maxHeight = 140.14.px
+                backgroundColor += (LinearGradient(0.0,0.0,0.0,140.0,false, CycleMethod.NO_CYCLE, Stop(0.0, Color.WHITE), Stop(1.0, Color.LIGHTGRAY)))
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditorStyles.kt]: parameters should be expanded out to their own lines for shorter line width

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditorStyles.kt [hunk 3]
```diff
@@ -52,14 +71,89 @@ class ProjectEditorStyles : Stylesheet() {
...
+                backgroundPosition += BackgroundPosition.CENTER
+                backgroundSize += BackgroundSize(0.0, 0.0,false, false, true, true)
+                fontSize = 12.px
+            }
+            versecardGraphic {
+                prefWidth = 160.0.px
+                prefHeight = 140.14.px
+                maxWidth = 160.0.px
+                maxHeight = 140.14.px
+                backgroundColor += (LinearGradient(0.0,0.0,0.0,140.0,false, CycleMethod.NO_CYCLE, Stop(0.0, Color.WHITE), Stop(1.0, Color.LIGHTGRAY)))
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditorStyles.kt]: line width again

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditorStyles.kt [hunk 4]
```diff
@@ -52,14 +71,89 @@ class ProjectEditorStyles : Stylesheet() {
...
+                borderRadius += box(5.0.px)
+                borderInsets += box(1.5.px)
+                backgroundImage += URI("/images/verse_image.png")
+                backgroundRepeat += Pair(BackgroundRepeat.NO_REPEAT, BackgroundRepeat.NO_REPEAT)
+                backgroundPosition += BackgroundPosition.CENTER
+                backgroundSize += BackgroundSize(0.0, 0.0,false, false, true, true)
+                fontSize = 16.px
+            }
+            cardLabel {
+                effect = DropShadow(25.0, 2.0,2.0, c("#FBFEFF"))
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditorStyles.kt]: same as above regarding color

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecthome/view/ProjectHomeStyles.kt [hunk 5]
```diff
@@ -50,33 +55,42 @@ class ProjectHomeStyles : Stylesheet() {
...
             }
         }
 
         projectCard {
-            prefWidth = 232.px
-            prefHeight = 300.px
-            backgroundColor += AppTheme.colors.cardBackground
+            prefWidth = 185.px
+            prefHeight = 225.px
+            backgroundColor += (LinearGradient(0.0,0.0,0.0,80.0,false, CycleMethod.NO_CYCLE, Stop(0.999, AppTheme.colors.appRed), Stop(1.0, Color.WHITE)))
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecthome/view/ProjectHomeStyles.kt]: line width

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                            cardNumber = label(chapter.titleKey)
+                        }
+                    }
+                    progressbar = progressbar(0.2) {
+
+                    }
+                }
+            }
+            cardButton = JFXButton("OPEN", MaterialIconView(MaterialIcon.ARROW_FORWARD).apply {
+                fill = c("#CC4141")
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt]: color variable

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            })
+            add(cardButton)
+        }
+    }
+}
+
+private fun percentComplete(collection: Collection) {
+}
+
+fun Pane.chaptercard(chapter: Collection? = null,
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt]: I'd like to know your reasoning behind making this an extension method vs subclassing pane

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/contentcard/ContentCard.kt [hunk 8]
```diff
@@ -38,8 +39,9 @@ class ContentCard(initialContent: Content? = null) : VBox() {
     }
 }
 
-fun contentcard(verse: Content, init: ContentCard.() -> Unit): ContentCard {
+fun Pane.contentcard(verse: Content, init: ContentCard.() -> Unit): ContentCard {
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/contentcard/ContentCard.kt]: see above

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt (hunk 7 above)
Comment (natibekele) [reply src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt]: I have no reasoning behind it. I was just following an implementation that Matthew had adopted previously

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/contentcard/ContentCard.kt (hunk 8 above)
Comment (natibekele) [reply src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/contentcard/ContentCard.kt]: I have no reasoning behind it. I was just following an implementation that Matthew had adopted previously

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditorStyles.kt [hunk 9]
```diff
@@ -52,12 +71,103 @@ class ProjectEditorStyles : Stylesheet() {
...
+                        true,
+                        true)
+                fontSize = 12.px
+            }
+            versecardGraphic {
+                prefWidth = 160.0.px
+                prefHeight = 140.14.px
+                maxWidth = 160.0.px
+                maxHeight = 140.14.px
+                backgroundColor += (LinearGradient(0.0,
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditorStyles.kt]:
```
LinearGradient(
                        0.0,
                        0.0,
                        0.0,
                        140.0,
                        false,
                        CycleMethod.NO_CYCLE,
                        Stop(0.0, Color.WHITE),
                        Stop(1.0, Color.LIGHTGRAY)
)
```

is there a reason for the beginning parenthesis?

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                chapterGraphic = vbox {
+                    alignment = Pos.CENTER
+                    stackpane {
+                        ellipse {
+                            centerX = 0.0
+                            centerY = 0.0
+                            radiusX = 60.0
+                            radiusY = 15.0
+                            fill = Color.WHITE
+                            effect = GaussianBlur(15.0)                        }
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt]: end } should be on its own line

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                        }
+                    }
+                    stackpane {
+                        ellipse {
+                            centerX = 50.0
+                            centerY = 0.0
+                            radiusX = 30.0
+                            radiusY = 30.0
+                            fill = Color.WHITE
+                            effect = GaussianBlur(15.0)                        }
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt]: end }

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+            add(cardButton)
+        }
+    }
+}
+
+private fun percentComplete(collection: Collection) {
+}
+
+fun Pane.chaptercard(chapter: Collection? = null,
+                     init: ChapterCard.() -> Unit = {}): ChapterCard {
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt:84]:
```
 fun Pane.chaptercard(
    chapter: Collection? = null,
    init: ChapterCard.() -> Unit = {}
) : ChapterCard {
```

https://kotlinlang.org/docs/reference/coding-conventions.html#class-header-formatting

(I know we're moving away from this method, but yeah)

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/projectcard/ProjectCard.kt [hunk 13]
```diff
@@ -8,8 +8,9 @@ import javafx.scene.Node
 import javafx.scene.control.Button
 import javafx.scene.control.Label
 import javafx.scene.layout.*
+import javafx.scene.paint.Color
 import org.wycliffeassociates.otter.common.data.model.Collection
-import org.wycliffeassociates.resourcecontainer.entity.Project
+//import org.wycliffeassociates.resourcecontainer.entity.Project
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/projectcard/ProjectCard.kt]: just delete the import, don't comment it out

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditor.kt [hunk 14]
```diff
@@ -64,23 +65,25 @@ class ProjectEditor : View() {
...
-                    textProperty().bind(viewModel.projectTitleProperty)
-                    addClass(ProjectEditorStyles.projectTitle)
-                }
-                childrenList = listview {
-                    items = viewModel.children
+                vgrow = Priority.ALWAYS
+                childrenList = projectnav(
+                        viewModel.projectProperty,
+                        viewModel.activeChildProperty,
+                        viewModel.activeContentProperty) {
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditor.kt]: passing the view model properties doesn't feel like the right way to solve this problem (looking to other UI controls in tfx, we bind properties there in a different way), though this problem may have come from trying to adhere to the constraint of running as an extension method, so the refactor may just naturally fix this

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditor.kt [hunk 15]
```diff
@@ -109,86 +105,64 @@ class ProjectEditor : View() {
...
+                            bindChildren(viewModel.children) {
+                                vbox {
+                                    alignment = Pos.CENTER
+                                    chaptercard(it) {
+                                        addClass(ProjectEditorStyles.collectionCard)
+                                        chapterGraphic.apply { addClass(ProjectEditorStyles.chaptercardGraphic) }
+                                        cardNumber.addClass(ProjectEditorStyles.cardNumber)
+                                        cardButton.addClass(ProjectEditorStyles.cardButton)
+                                        cardBackground.addClass(ProjectEditorStyles.cardBackground)
+                                        progressbar.addClass(ProjectEditorStyles.cardProgressbar)
```
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditor.kt:124]: similar to above, setting styles here seems awkward and they probably should be style sheets for cards themselves, not this page (unless the page is explicitly overriding defaults?)?

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditor.kt (hunk 14 above)
Comment (natibekele) [reply src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditor.kt]: my thinking was following how tfx handles lables. the labels can just be handed some type of observable value and the label displays that. my thinking was if we hand the certain properties to the projectnav then the nav will always display them as they change. idk

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditor.kt (hunk 15 above)
Comment (natibekele) [reply src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/projecteditor/view/ProjectEditor.kt:124]: yes. the idea is that we are overriding the default styles for the cards here
will be important when we have to change the style/color of the cards depending on the most recently active project/chapter/verse + completed projects chapters and verses

File: src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+    var card: VBox by singleAssign()
+    var cardLabel: Label by singleAssign()
+    var cardNumber: Label by singleAssign()
+    var cardButton: Button by singleAssign()
+    var chapterGraphic: Node = StackPane()
+    var cardBackground: VBox by singleAssign()
+    var progressbar: ProgressBar by singleAssign()
+    private val defaultFill = c("#CC4141")
+    init {
```
Comment (KJoslyn) [inline src/main/kotlin/org/wycliffeassociates/otter/jvm/app/widgets/chaptercard/ChapterCard.kt:27]: Picky, but empty line above init?

**Pull request closed without merging**